### PR TITLE
Campaign calendar

### PR DESCRIPTION
**Pull request by pekc83**
Closes issue #1

### CODE REVIEW

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.html [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        <th scope="col" style="width: 9%">Start Date</th>
+        <th scope="col" style="width: 9%">End Date</th>
+        <th scope="col" style="width: 10%">Assigned</th>
+      </tr>
+    </thead>
+    <tbody>
+    <ng-container *ngFor="let campaign of campaigns; trackBy: trackByFn">
+      <tr (dblclick)="editCampaign(campaign)">
+        <td>
+          <ng-container  *ngIf="!readonly">
```
Comment (integry) [inline src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.html]: Why not to hide the entire cell? You would also have to hide the header cell, but I think it's better than to have an empty gap on the left.

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import { CampaignDeletionConfirmationComponent } from './campaign-deletion-confirmation/campaign-deletion-confirmation';
+
+@Component({
+  selector: 'campaign-calendar',
+  templateUrl: './campaign-calendar.component.html',
+  styleUrls: ['./campaign-calendar.component.sass']
+})
+export class CampaignCalendarComponent implements OnInit {
+  @Input() readonly = false;
+  @Output() campaignsChange: EventEmitter<Campaign[]> = new EventEmitter<Campaign[]>();
```
Comment (integry) [inline src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts]: I think no need to specify the type as it can be inferred from the value.

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  @Input() campaigns: Campaign[] = [];
+  months: string[];
+  campaignGraphs: { [key: string]: Campaign[] } = {};
+  years: string[] = [];
+  colors: string[] = websafeColors.slice();
```
Comment (integry) [inline src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts]: Same as above

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        this.updateCampaigns();
+      })
+      .catch(this.doNothing);
+  }
+
+  editCampaign(campaign: Campaign) {
+    if (this.readonly) {
+      return;
+    }
+    const data = this.campaigns.find((item: Campaign) => item.id === campaign.id);
```
Comment (integry) [inline src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts]: No need to specify `item` type, as Typescript already knows it from `this.campaigns`

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    const offsetDays = start.dayOfYear() - 1;
+    const persentageOffset = offsetDays / this.getDaysInYear(start);
+    const borderWidth = 1;
+    const monthsCellWidth = table.offsetWidth - 2 * borderWidth - tableLabel.offsetWidth;
+    const offset = Math.round(monthsCellWidth * persentageOffset);
+    return tableLabel.offsetWidth + offset;
+  }
+
+  private splitCampaignsByYear(campaigns: Campaign[]) {
+    campaigns.forEach((campaign: Campaign) => {
```
Comment (integry) [inline src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts]: Could this be rewritten as reduce?

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+      }
+      if (start.year() !== end.year()) {
+        this.splitCampaignsByYear(this.splitCampaignByYear(campaign));
+      } else {
+        this.campaignGraphs[start.year()].push(campaign);
+      }
+    });
+  }
+
+  private splitCampaignByYear(campaign: Campaign): Campaign[] {
```
Comment (integry) [inline src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts]: I think you can pass all campaigns through this function without pre-checking the dates. In case the campaign fits inside one year, it returns an array or object with one element, otherwise it's 2 elements or more. It would make it easier to change forEach to reduce above as well.

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    );
+    this.updateCampaigns();
+  }
+
+  openModal() {
+    this.modalService
+      .open(ModalAddCampaignComponent)
+      .result
+      .then(this.addCampaign.bind(this))
+      .catch(this.doNothing);
```
Comment (integry) [inline src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts]: Why is this necessary?

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+  getCampaignOffset(table: HTMLElement, tableLabel: HTMLElement, campaign: Campaign): number {
+    const start = moment(campaign.startDate, this.dateFormat);
+    const offsetDays = start.dayOfYear() - 1;
+    const persentageOffset = offsetDays / this.getDaysInYear(start);
+    const borderWidth = 1;
+    const monthsCellWidth = table.offsetWidth - 2 * borderWidth - tableLabel.offsetWidth;
+    const offset = Math.round(monthsCellWidth * persentageOffset);
+    return tableLabel.offsetWidth + offset;
+  }
```
Comment (integry) [inline src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts]: Could the sizes be defined by percentages instead of having to access the elements directly?

**Comment by integry** [comment]
Got an error when trying to modify one existing entry:

image

File: src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts (hunk 7 above)
Comment (pekc83) [reply src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template/campaign-calendar/campaign-calendar.component.ts]: It's necessary to avoid appearing of error in browser console on modal dismiss by clicking cancel or `x` button on modal.

File: src/app/module-builder/module-editor/module-editor.component.ts [hunk 9]
```diff
@@ -9,6 +9,7 @@ import { StepLinkEditorComponent } from './step-link-editor/step-link-editor.com
 import { Subscription } from 'rxjs';
 import { switchMap, catchError } from 'rxjs/operators';
 import { ToastrService } from 'ngx-toastr';
+import { CampaignCalendarTemplateData } from 'src/app/module-viewer/riverside-step-template/templates/campaign-calendar-template';
```
[the remark this answers is not in the document]
Comment (pekc83) [reply src/app/module-builder/module-editor/module-editor.component.ts]: Yes it was.

**Comment by pekc83** [comment]
> Got an error when trying to modify one existing entry:
> 
> image

That's very strange, looks like campaign came from backend without `Theme`. However, I've fixed it and this error will not appear any more.
